### PR TITLE
Added CNN global pooling layer

### PR DESCRIPTION
**Pull request by entron**

### CODE REVIEW

**Comment by fchollet** [comment]
One issue: this would break model serialization / deserialization (`pooling_function`).

**Comment by entron** [comment]
Sorry I don't understand what do you mean by serialization. Could you explain a bit more?

**Comment by fchollet** [comment]
You have the ability to save and load model architectures via `model.to_json()` / `model.to_yaml()` and `model_from_json(string)` / model_from_yaml(string)`. That's model seralization / deserialization. 

The model reconstruction step would break here.

**Comment by entron** [comment]
I didn't know there are such useful methods. I just tried it but even with a model without using this layer I still got an error:

```
Traceback (most recent call last):
  File "train_keras.py", line 46, in <module>
    model.to_yaml()
  File "/home/cheng/keras/keras/models.py", line 317, in to_yaml
    return yaml.dump(config)
  File "/usr/local/lib/python3.4/dist-packages/yaml/__init__.py", line 200, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "/usr/local/lib/python3.4/dist-packages/yaml/__init__.py", line 188, in dump_all
    dumper.represent(data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 26, in represent
    node = self.represent_data(data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 203, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 116, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 195, in represent_list
    return self.represent_sequence('tag:yaml.org,2002:seq', data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 91, in represent_sequence
    node_item = self.represent_data(item)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 203, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 116, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 51, in represent_data
    node = self.yaml_multi_representers[data_type](self, data)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 342, in represent_object
    return self.represent_sequence(tag+function_name, args)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 91, in represent_sequence
    node_item = self.represent_data(item)
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 33, in represent_data
    if self.ignore_aliases(data):
  File "/usr/local/lib/python3.4/dist-packages/yaml/representer.py", line 135, in ignore_aliases
    if data in [None, ()]:
TypeError: data type not understood
```

Do you know where could be wrong?

**Comment by fchollet** [comment]
You are serializing data that is not serializable. What is your model?

**Comment by entron** [comment]
I just used the example VGG model of keras as this:

```
def Keras_Example_VGG_32(n_classes):
    model = Sequential()
    model.add(Convolution2D(32, 3, 3, 3, border_mode='full')) 
    model.add(Activation('relu'))
    model.add(Convolution2D(32, 32, 3, 3))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(poolsize=(2, 2)))
    model.add(Dropout(0.25))

    model.add(Convolution2D(64, 32, 3, 3, border_mode='full')) 
    model.add(Activation('relu'))
    model.add(Convolution2D(64, 64, 3, 3)) 
    model.add(Activation('relu'))
    model.add(MaxPooling2D(poolsize=(2, 2)))
    model.add(Dropout(0.25))

    model.add(Flatten())
    model.add(Dense(64 * 8 * 8, 256))
    model.add(Activation('relu'))
    model.add(Dropout(0.5))

    model.add(Dense(256, n_classes))
    model.add(Activation('softmax'))
    return model
```

**Comment by fchollet** [comment]
Yaml and Json serialization both work fine on this example (just tested it here).

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -245,3 +245,17 @@ def get_output(self, train):
...
+        super(GlobalPooling, self).__init__()
+        self.pooling_function = pooling_function
+
+    def get_output(self, train):
+        X = self.get_input(train)
+        return self.pooling_function(X.flatten(3), axis=-1)
+
+    def get_config(self):
+        return {"name": self.__class__.__name__,
+                "pooling_function}": self.pooling_function.__name__}
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Syntax error here (`}`)

Comment (fchollet) [inline keras/layers/convolutional.py]: Also: this layer seems to assume a Tensor4 as input but does not explicitly define a `self.input` attribute. Also if it is 2D-specific that should appear in the name (`GlobalPooling2D`).

**Comment by entron** [comment]
I have fixed the problems you pointed out. However, the to_yaml() still not working for both model on my computer. With pdb I printed "config":

```
{'name': 'Sequential', 'layers': [{'b_regularizer': None, 'subsample': (1, 1), 'activation': 'linear', 'stack_size': 3, 'init': 'glorot_uniform', 'W_constraint': None, 'W_regularizer': None, 'b_constraint': None, 'name': 'Convolution2D', 'border_mode': 'full', 'nb_col': 3, 'nb_row': 3, 'nb_filter': 32, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'relu', 'beta': 0.1}, {'b_regularizer': None, 'subsample': (1, 1), 'activation': 'linear', 'stack_size': 32, 'init': 'glorot_uniform', 'W_constraint': None, 'W_regularizer': None, 'b_constraint': None, 'name': 'Convolution2D', 'border_mode': 'valid', 'nb_col': 3, 'nb_row': 3, 'nb_filter': 32, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'relu', 'beta': 0.1}, {'name': 'MaxPooling2D', 'ignore_border': True, 'stride': None, 'poolsize': (2, 2)}, {'name': 'Dropout', 'p': 0.25}, {'b_regularizer': None, 'subsample': (1, 1), 'activation': 'linear', 'stack_size': 32, 'init': 'glorot_uniform', 'W_constraint': None, 'W_regularizer': None, 'b_constraint': None, 'name': 'Convolution2D', 'border_mode': 'full', 'nb_col': 3, 'nb_row': 3, 'nb_filter': 64, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'relu', 'beta': 0.1}, {'b_regularizer': None, 'subsample': (1, 1), 'activation': 'linear', 'stack_size': 64, 'init': 'glorot_uniform', 'W_constraint': None, 'W_regularizer': None, 'b_constraint': None, 'name': 'Convolution2D', 'border_mode': 'valid', 'nb_col': 3, 'nb_row': 3, 'nb_filter': 64, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'relu', 'beta': 0.1}, {'name': 'MaxPooling2D', 'ignore_border': True, 'stride': None, 'poolsize': (2, 2)}, {'name': 'Dropout', 'p': 0.25}, {'name': 'Flatten'}, {'name': 'Dense', 'b_regularizer': None, 'W_regularizer': None, 'input_dim': 4096, 'activation': 'linear', 'b_constraint': None, 'init': 'glorot_uniform', 'output_dim': 256, 'W_constraint': None, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'relu', 'beta': 0.1}, {'name': 'Dropout', 'p': 0.5}, {'name': 'Dense', 'b_regularizer': None, 'W_regularizer': None, 'input_dim': 256, 'activation': 'linear', 'b_constraint': None, 'init': 'glorot_uniform', 'output_dim': 10, 'W_constraint': None, 'activity_regularizer': None}, {'name': 'Activation', 'target': 0, 'activation': 'softmax', 'beta': 0.1}]}
```
and the "data" variable where the error occurs: `dtype('int64')`

Do you know where could be my problem?

**Comment by maxpumperla** [comment]
@entron I think the problem is that you put `self.pooling_function.__name__` into the config, which is something like `T.mean`. Your constructor takes a string, not a function, so you should rather put `average` there, instead. Hope this helps.

**Comment by EderSantana** [comment]
@entron could you consider restarting your PR with the new backend? Your layer will look like this:
```python
class GlobalPooling2D(Layer):
    """
    By @entron
    Borrowed and modified from here: https://github.com/fchollet/keras/pull/522
    """
    def __init__(self, pooling_function='average'):
        super(GlobalPooling2D, self).__init__()
        if pooling_function not in {'average', 'max'}:
            raise Exception('Invalid pooling function for GlobalPooling2D:', pooling_function)
        if pooling_function == 'average':
            self.pooling_function = K.mean
        else:
            self.pooling_function = K.max
        self.input = K.placeholder(ndim=4)

    def get_output(self, train):
        X = self.get_input(train)
        return self.pooling_function(self.pooling_function(X, axis=-1), axis=-1)

    def get_config(self):
        return {"name": self.__class__.__name__,
                "pooling_function": self.pooling_function.__name__}
```

Please also add a test like the following one to test_convolutional.py:

```python
from numpy.testing import asser_allclose
def test_global_pooling():
    layer = GlobalPooling2D()
    X = K.placeholder(ndim=4)
    layer.input = X
    Y = layer.get_output()
    F = function([X], Y)
    
    x = np.ones(3, 1, 4, 4)
    t  = x.mean(axis=(2, 3))
    y = F(x)
    np.testing.assert_allclose(y, t)
```

**Comment by EderSantana** [comment]
Let me know if you prefer me to do it myself.

**Comment by entron** [comment]
hi @EderSantana, thanks a lot! As I am quite occupied recently, so it would be great if you could create a new PR!

**Comment by maxpumperla** [comment]
Hey @EderSantana, I'm actually working on this as we speak, so if you don't mind I could take over this one and you could focus on your awesome VAEs instead. What do you think?

**Comment by EderSantana** [comment]
@maxpumperla Thanks!!! Please go ahead. I'll need that for the VAE :) that and `__call__` #1162

**Comment by xFlux** [comment]
Hi @EderSantana @maxpumperla . Do we have now a GlobalPooling Layer available in the current Keras codebase? I can't find it. Thank you for your help!

**Comment by maxpumperla** [comment]
Hi @xFlux, you find them here:

https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py

The PR relating to it was this one:
https://github.com/fchollet/keras/pull/1202

Since then, 3D Pooling has made it into the fold. I hope this helps.

**Comment by xFlux** [comment]
Hi @maxpumperla. thanks for your reply. i was having a look at convolutional.py, but it seems that all layers require the explicit definition of the "pool_size". Can I somehow use this to build a GlobalPooling layer that pools all the pixels of the previous layer, independent of the input image size and thus of the size of the previous layer?

**Pull request closed without merging**